syntax = "proto3";
option java_multiple_files = true;
package org.bank.grpc.Bank;

message Statement {
  int32 accountId = 1;
  string type = 2;
  string message = 3;
  string date = 4;
}

message Account {
  int32 id = 1;
  optional int32 pin = 2;
  string name = 3;
  double balance = 4;
  double limit = 5;
  string date = 6;
}

message Ok {
  bool ok = 1;
  optional string message = 2;
}

message GenericRequest {
  int32 id = 1;
}

message AuthRequest {
  int32 id = 1;
  int32 pin = 2;
}

message RegisterRequest {
  int32 id = 1;
  int32 pin = 2;
  string name = 3;
}

message DepositRequest {
  int32 id = 1;
  double amount = 2;
}
message WithdrawRequest {
  int32 id = 1;
  double amount = 2;
}

message TransferRequest {
  int32 id_from = 1;
  int32 id_to = 2;
  float amount = 3;
}

message InfoResponse {
  Account account = 1;
  Ok ok = 2;
}


message BalanceResponse {
  double balance = 1;
  Ok ok = 2;
}

message StatementResponse {
  repeated Statement statement = 1;
  Ok ok = 2;
}


message LoginResponse {
  int32 id = 1;
  string name = 2;
  Ok ok = 3;
}

message RegisterResponse {
  int32 id = 1;
  string name = 2;
  Ok ok = 3;
}

message DepositResponse {
  double amount = 1;
  Ok ok = 2;
}

message WithdrawResponse {
  double amount = 1;
  Ok ok = 2;
}

message TransferResponse {
  double amount = 1;
  Ok ok = 2;
}

service Bank {
  rpc info(GenericRequest) returns (InfoResponse) {}
  rpc statements(GenericRequest) returns (StatementResponse) {}
  rpc balance(GenericRequest) returns (BalanceResponse) {}
  rpc login(AuthRequest) returns (LoginResponse) {}
  rpc register(RegisterRequest) returns (RegisterResponse) {}
  rpc deposit(DepositRequest) returns (DepositResponse) {}
  rpc withdraw(WithdrawRequest) returns(WithdrawResponse) {}
  rpc transfer(TransferRequest) returns (TransferResponse) {}
}

